credoai/utils/model_utils.py

```diff
@@ -125,3 +170,165 @@ def validate_dummy(model_like, _):
             validate_sklearn_like(model_like.model_like, tmp_model_info)
         else:
             raise
+
+
+def reg_handle_torch(model_like):
+    pred_func = getattr(model_like, "forward", None)
+    if pred_func is None:
+        raise ValueError("Model should have a `forward` method to perform predictions.")
+
+    return pred_func
+
+
+def clf_handle_keras(model_like):
+    predict_obj = None
+    predict_proba_obj = None
+    clf_type = "BINARY_CLASSIFICATION"
+
+    pred_func = getattr(model_like, "predict", getattr(model_like, "call"))
+
+    if model_like.layers[-1].output_shape == (None, 1):
+        # Assumes sigmoid -> probabilities need to be rounded
+        predict_obj = lambda x: pred_func(x).round()
+        # Single-output sigmoid is binary by definition
+        clf_type = "BINARY_CLASSIFICATION"
+    else:
+        # Assumes softmax -> probabilities need to be argmaxed
+        predict_obj = lambda x: np.argmax(pred_func(x), axis=1)
+        if model_like.layers[-1].output_shape[1] == 2:
+            clf_type = "BINARY_CLASSIFICATION"
+        else:
+            clf_type = "MULTICLASS_CLASSIFICATION"
+
+    if model_like.layers[-1].output_shape == (None, 2):
+        predict_proba_obj = lambda x: pred_func(x)[:, 1]
+    elif (
+        len(model_like.layers[-1].output_shape) == 2
+        and model_like.layers[-1].output_shape[1] == 1
+    ):
+        # Sigmoid -> needs to be (n_samples, ) to work with sklearn metrics
+        predict_proba_obj = lambda x: np.reshape(pred_func(x), (-1, 1))
+    elif (
+        len(model_like.layers[-1].output_shape) == 2
+        and model_like.layers[-1].output_shape[1] > 2
+    ):
+        predict_proba_obj = pred_func
+    else:
+        pass
+        # predict_proba is not valid (for now) --> this would correspond to multi-dimensional outputs or something similarly weird
+
+    return predict_obj, predict_proba_obj, clf_type
+
+
+def clf_handle_torch(model_like):
```

Comment:
I like how you separated handle functions. Can you add a docstring that gives the gist of what is going on?


```
### This is what Claude suggests 😛 
Creates a standardized interface for making inferences from a PyTorch classification model with different activations and output types.

- It checks that the model passed in has a forward method for making predictions. If not, it raises an error.
- It gets the output activation used by the model, either "sigmoid" or "softmax". If an invalid activation is specified, it defaults to "sigmoid".
- It performs a dummy prediction with the model to get the shape of the output. It uses this to determine if the model is a binary or multi-class classifier.
- It creates a predict function that performs inferences on either a single input or a batch of inputs. For sigmoid activation, it returns rounded predictions. For softmax, it returns the predicted class label.
- It creates a predict_proba function to return raw prediction probabilities. For sigmoid, it returns the logistic prediction. For softmax, it returns the probability of the positive class for binary classification or the full probability distribution for multi-class classification.
- It returns the predict and predict_proba functions along with a string specifying the classifier type (binary or multi-class).
- These returned functions can then be used to easily make predictions and get probabilities from the PyTorch model. The functions handle the differences between sigmoid and softmax activations and between binary and multi-class classification models.
"""
```